// trophy-common-utils/DatabaseTableFields.mjs
export const WEIGHT_ATTRIBUTE_VALUE_COLUMN = 'weight_attribute_value';
export const WEIGHT_ATTRIBUTE_UNIT_VALUE_COLUMN = 'weight_attribute_unit_value';
export const WEIGHT_ATTRIBUTE_UNIT_ID_COLUMN = 'weight_attribute_unit_id';
export const WEIGHT_ATTRIBUTE_ID_COLUMN = 'weight_attribute_id';
export const VARCHAR_COLUMN = 'varchar_column';
export const USER_ID_COLUMN = 'user_id';
export const USER_DEFINED_UNIT_ID_COLUMN = 'user_defined_unit_id';
export const USER_DEFINED_ATTRIBUTE_VALUE_COLUMN = 'user_defined_attribute_value';
export const USER_DEFINED_ATTRIBUTE_UNIT_VALUE_COLUMN = 'user_defined_attribute_unit_value';
export const USER_DEFINED_ATTRIBUTE_UNIT_ID_COLUMN = 'user_defined_attribute_unit_id';
export const USER_DEFINED_ATTRIBUTE_NAME_COLUMN = 'user_defined_attribute_name';
export const USER_DEFINED_ATTRIBUTE_ID_COLUMN = 'user_defined_attribute_id';
export const TIME_ATTRIBUTE_SECONDS_VALUE_COLUMN = 'time_attribute_seconds_value';
export const TIME_ATTRIBUTE_ID_COLUMN = 'time_attribute_id';
export const TEXT_COLUMN = 'text_column';
export const TEST_ID_COLUMN = 'test_id';
export const SETS_ATTRIBUTE_VALUE_COLUMN = 'sets_attribute_value';
export const SETS_ATTRIBUTE_ID_COLUMN = 'sets_attribute_id';
export const REPS_ATTRIBUTE_VALUE_COLUMN = 'reps_attribute_value';
export const REPS_ATTRIBUTE_ID_COLUMN = 'reps_attribute_id';
export const LEVEL_ATTRIBUTE_VALUE_ID_COLUMN = 'level_attribute_value_id';
export const LEVEL_ATTRIBUTE_VALUE_COLUMN = 'level_attribute_value';
export const LEVEL_ATTRIBUTE_ID_COLUMN = 'level_attribute_id';
export const INTENSITY_ATTRIBUTE_VALUE_ID_COLUMN = 'intensity_attribute_value_id';
export const INTENSITY_ATTRIBUTE_VALUE_COLUMN = 'intensity_attribute_value';
export const INTENSITY_ATTRIBUTE_UNIT_ID_COLUMN = 'intensity_attribute_unit_id';
export const INTENSITY_ATTRIBUTE_ID_COLUMN = 'intensity_attribute_id';
export const INT_COLUMN = 'int_column';
export const EXERCISE_TYPES_VALUE_COLUMN = 'exercise_types_value';
export const EXERCISE_TYPE_VALUE_COLUMN = 'exercise_type_value';
export const EXERCISE_TYPES_ID_COLUMN = 'exercise_types_value';
export const EXERCISE_TYPE_ID_COLUMN = 'exercise_type_id';
export const EXERCISE_TRACKING_STATUS_COLUMN = 'exercise_tracking_status';
export const EXERCISE_NOTES_COLUMN = "exercise_notes";
// export const EXERCISE_NOTES_VALUE_COLUMN = 'exercise_notes_value';
// export const EXERCISE_NOTES_ID_COLUMN = 'exercise_notes_id';
export const EXERCISE_NAME_COLUMN = 'exercise_name';
export const EXERCISE_ID_COLUMN = 'exercise_id';
export const EXERCISE_DATETIME_CREATED_COLUMN = 'exercise_datetime_created';
export const EXERCISE_ATTRIBUTE_ID_COLUMN = 'exercise_attribute_id';
export const EXERCISE_ATTRIBUTES_ID_COLUMN = 'exercise_attributes_id';
export const DISTANCE_ATTRIBUTE_VALUE_COLUMN = 'distance_attribute_value';
export const DISTANCE_ATTRIBUTE_UNIT_VALUE_COLUMN = 'distance_attribute_unit_value';
export const DISTANCE_ATTRIBUTE_UNIT_ID_COLUMN = 'distance_attribute_unit_id';
export const DISTANCE_ATTRIBUTE_ID_COLUMN = 'distance_attribute_id';
export const DECIMAL_COLUMN = 'decimal_column';
export const DATETIME_COLUMN = 'datetime_column';
export const DATE_COLUMN = 'date_column';
export const BOOLEAN_COLUMN = 'boolean_column';
