// trophy-common-utils/DatabaseTables.mjs
export const DISTANCE_ATTRIBUTE_UNITS_TABLE = "distance_attribute_units";
export const DISTANCE_ATTRIBUTES_TABLE = "distance_attributes";
export const EXERCISE_ATTRIBUTES_TABLE = "exercise_attributes";
export const EXERCISE_NOTES_TABLE = "exercise_notes";
export const EXERCISE_TYPES_TABLE = "exercise_types";
export const EXERCISES_TABLE = "exercises";
export const INTENSITY_ATTRIBUTE_VALUES_TABLE = "intensity_attribute_values";
export const INTENSITY_ATTRIBUTES_TABLE = "intensity_attributes";
export const LEVEL_ATTRIBUTE_VALUES_TABLE = "level_attribute_values";
export const LEVEL_ATTRIBUTES_TABLE = "level_attributes";
export const REPS_ATTRIBUTES_TABLE = "reps_attributes";
export const SETS_ATTRIBUTES_TABLE = "sets_attributes";
export const TEST_TABLE_TABLE = "test_table";
export const TIME_ATTRIBUTES_TABLE = "time_attributes";
export const USER_DEFINED_ATTRIBUTE_UNITS_TABLE = "user_defined_attribute_units";
export const USER_DEFINED_ATTRIBUTES_TABLE = "user_defined_attributes";
export const WEIGHT_ATTRIBUTE_UNITS_TABLE = "weight_attribute_units";
export const WEIGHT_ATTRIBUTES_TABLE = "weight_attributes";
