// trophy-common-utils/Errors.mjs
export const ERROR = {

    ERROR_PARAMETER_UNDEFINED: "Expected Parameter to be defined but is undefined",

    ERROR_RDS_CONNECTION_FAILED: "Connection to the RDS instance failed",

    ERROR_ENV_VARIABLE_NOT_FOUND: "Environment Variable not found",

    ERROR_JSON_INVALID_JSON_OBJECT: "JSON does not meet required formatting as a JSON object",
    ERROR_JSON_INVALID_PROPERTIES: "JSON properties do not meet requirements. 1+ undefined properties",
    ERROR_JSON_INVALID_PROPERTY_TYPES: "JSON property values do not meet requirements. 1+ are incorrect data type",
    ERROR_JSON_INVALID_EXERCISE_FIELD_VALUE: "Exercise JSON has at least 1 undefined or non-null value in a required field",

    ERROR_SECRETS_MANAGER_COULD_NOT_GET_SECRET_VALUE: "Secrets Manager failed performing getSecretValue() call",

    ERROR_EXERCISE_PARAMETER_NULL_UNDEFINED: "Null or Undefined parameter found when performing Create operation for Exercise",
    ERROR_EXERCISE_PARAMATERS_INVALID: "Exercise contains invalid parameters",

    ERROR_EXERCISE_COULD_NOT_RETRIEVE: "Retrieve operation failed for Exercise",

    ERROR_EXERCISE_ATTRIBUTES_INVALID: "Exercise Attributes do not meet requirements",
    
    ERROR_TRANSLATE_EXERCISE_SQL_TO_JSON: "Translating Exercise SQL Row to JSON object failed",

    ERROR_CREATE_EXERCISE_FAILED: "Create database-operation failed for Exercise",
    ERROR_RETRIEVE_EXERCISE_FAILED: "Retrieve database-operation failed for Exercise",
    ERROR_UPDATE_EXERCISE_FAILED: "Update database-operation failed for Exercise",
    ERROR_DELETE_EXERCISE_FAILED: "Delete database-operation failed for Exercise",

    ERROR_EXERCISE_ATTRIBUTE_DELETE_FAILED: "Delete database-operation failed for Exercise Attribute",

    ERROR_EXERCISE_JSON_INVALID_FORMAT: "Exercise JSON does not meet formatting requirements",
    ERROR_EXERCISE_ATTRIBUTE_JSON_INVALID_FORMAT: "Exercise Attribute JSON does not meet formatting requirements"
}
