// lambda-functions/gptProcessUserExercise.mjs
import { validatePathParametersUserIdWithEvent } from "../trophy-common-utils/PathParameterValidator.mjs"
import { callGptApi, processGptOutput, invokePutUserExercise } from "../trophy-gpt-processing/index.mjs";

export const handler = async function (event, context) {

    try {
        // Validate we can access the user id property the user id from the path
        validatePathParametersUserIdWithEvent(event);

        // Get the GPT API response + Process the response
        const gptResponse = await callGptApi(userInput);
        const processedJson = processGptOutput(gptResponse);

        // Invoke putUserExercise Lambda function
        const lambdaResponse = await invokePutUserExercise(userId, processedJson);
        console.log(lambdaResponse);
        return {
            statusCode: 200,
            body: JSON.stringify({
                message: 'Input processed successfully',
                data: lambdaResponse
            })
        };

    } catch(error) {
        return {
            statusCode: 500,
            body: JSON.stringify({
                event: event,
                message: error.message,
                stack: error.stack
            })
        };
    } finally {

    }

}
